engine/src/main/java/org/hibernate/search/engine/environment/thread/spi/BatchingThreadContext.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.environment.thread.spi;
+
+import org.hibernate.search.util.common.annotation.Incubating;
+
+/**
+ * A thread context that can be used to manipulate the execution environment, for example,
+ * to initialize {@link ThreadLocal ThreadLocals}.
+ * <p>
+ * See {@code java.util.concurrent.ThreadPoolExecutor#beforeExecute(Thread, Runnable)}.
+ * See {@code java.util.concurrent.ThreadPoolExecutor#afterExecute(Thread, Runnable)}.
+ */
+@Incubating
+public interface BatchingThreadContext {
+
+	BatchingThreadContext DO_NOTHING = new BatchingThreadContext() {
+		@Override
+		public void beforeExecution(Thread t, Runnable r) {
+			// Do nothing.
+		}
+
+		@Override
+		public void afterExecution(Runnable r, Throwable t) {
+			// Do nothing.
+		}
+	};
+
+	/**
+	 * Method is going to be invoked prior to executing the {@link Runnable} in the given thread.
+	 */
+	void beforeExecution(Thread t, Runnable r);
```

Comment:
It's probably not that useful to pass the `Thread` because it can be accessed through `Thread.currentThread()` (and I'm not sure why one would do that).

I doubt it's useful to pass the `Runnable`, and that could actually be harmful (what if the user calls `r.run()`?).

Reply:
right ... that's just what the `ThreadPoolExecutor` uses for the arguments ...

> what if the user calls r.run()?

well ... that would be a shame 😄